drivers/block-nbd.c

```diff
@@ -554,68 +563,68 @@ negotiate_client_newstyle_options(int sock, td_driver_t *driver)
 	new_option.optlen = htobe32(sizeof(exportname));
 	new_option.option = htobe32( NBD_OPT_EXPORT_NAME);
 
+	struct nbd_export_name_option_reply handshake_finish;
+	static const size_t NO_ZERO_HANDSHAKE_FINISH_SIZE = 10;
+
 	/* Send EXPORTNAME_NAME option request */
 	rc = send_fully_or_fail(sock, &new_option, sizeof(new_option));
 	if (rc < 0)
 	{
 		ERROR("Failed to send options to sock");
-		close(sock);
-		return -1;
+		goto errout;
 	}
 	/* Send exportname name */
 	rc = send_fully_or_fail(sock, exportname, sizeof(exportname));
 	if (rc < 0)
 	{
 		ERROR("Failed to send export name to sock");
-		close(sock);
-		return -1;
+		goto errout;
 	}
 
 	/* Collect the results in the handshake finished */
-	struct nbd_export_name_option_reply handshake_finish;
-	static const size_t NO_ZERO_HANDSHAKE_FINISH_SIZE = 10;
-	rc = recv_fully_or_fail(sock, &handshake_finish, NO_ZERO_HANDSHAKE_FINISH_SIZE);
+	rc = tdnbd_wait_recv(sock, &handshake_finish, NO_ZERO_HANDSHAKE_FINISH_SIZE, 0);
```

Comment:
The code used to use recv_fully_or_fail, so it looks like it's assuming that if recv doesn't fail then it will always read the full amount. Elsewhere the same assumption is being made. Is there something at a more global level that's making sure this is OK?

Reply:
I would like to draw your attention to a few lines below, where
```
rc = fcntl(sock, F_SETFL, O_NONBLOCK);
```
happens. The socket is still blocking at this point, so if we get to the recv(), it _will_ receive NO_ZERO_HANDSHAKE_FINISH_SIZE octets. ```tnbd_wait_recv()``` ensures the socket has become readable before it attempts to read, which ensures that the ```recv()``` will not block forever.

At this point anything other than a full read indicates that Something Bad happened to the peer.